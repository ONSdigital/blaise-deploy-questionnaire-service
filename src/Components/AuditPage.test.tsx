import React from "react";
import {render, waitFor, cleanup, fireEvent, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import flushPromises, {mock_server_request_Return_JSON} from "../tests/utils";
import {act} from "react-dom/test-utils";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import AuditPage from "./AuditPage";


const auditLogsList = [
    {id: "602fb3250003c61e92b25da0", timestamp: "Fri Feb 19 2021 12:46:29 GMT+0000 (Greenwich Mean Time)", message: "Successfully uninstalled questionnaire OPN2012K", severity: "INFO"}
];

const auditLogsList2 = [
    {id: "602fb3250003c61e92b25da0", timestamp: "Fri Feb 19 2021 12:47:29 GMT+0000 (Greenwich Mean Time)", message: "Successfully installed questionnaire OPN2012K", severity: "INFO"}
];

describe("Audit Logs page", () => {

    beforeEach(() => {
        mock_server_request_Return_JSON(200, auditLogsList);
    });

    it("view Audit Logs page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/Questionnaire deployment history/i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

        await waitFor(() => {
            expect(getByText(/Questionnaire deployment history/i)).toBeDefined();
            expect(getByText(/Successfully uninstalled questionnaire OPN2012K/i)).toBeDefined();
            expect(getByText(/19\/02\/2021 12:46:29/i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });
    });

    it("should refresh the list when you press the Reload logs button", async () => {
        const history = createMemoryHistory();
        const {getByText} = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(getByText(/Successfully uninstalled questionnaire OPN2012K/i)).toBeDefined();
            expect(getByText(/19\/02\/2021 12:46:29/i)).toBeDefined();
        });

        mock_server_request_Return_JSON(200, auditLogsList2);

        await fireEvent.click(screen.getByText("Reload"));

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(getByText(/Successfully installed questionnaire OPN2012K/i)).toBeDefined();
            expect(getByText(/19\/02\/2021 12:47:29/i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns a 500 status", () => {

    beforeEach(() => {
        mock_server_request_Return_JSON(500, []);
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();


        await waitFor(() => {
            expect(getByText(/Unable to load deployment history./i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns malformed json", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, {text: "Hello"});
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();


        await waitFor(() => {
            expect(getByText(/Unable to load deployment history./i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns an empty list", () => {

    beforeEach(() => {
        mock_server_request_Return_JSON(200, []);
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <AuditPage/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();


        await waitFor(() => {
            expect(getByText(/No recent deployment history found./i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
