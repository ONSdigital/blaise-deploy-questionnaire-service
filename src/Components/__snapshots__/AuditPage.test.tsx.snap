// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit Logs page view Audit Logs page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="breadcrumb"
      >
        <ol
          class="breadcrumb__items u-fs-s"
        >
          <li
            class="breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="breadcrumb__link"
              href="/"
            >
              Home
            </a>
            <svg
              class="svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <h1
          class="u-mb-l"
        >
          Questionnaire deployment history
        </h1>
        <button
          class="btn btn--small "
          data-testid="button"
          type="button"
        >
          <span
            class="btn__inner"
          >
            Reload
          </span>
        </button>
        <div
          class="elementToFadeIn"
        >
          <table
            class="table"
            id="audit-table"
          >
            <thead
              class="table__head"
            >
              <tr
                class="table__row"
              >
                <th
                  class="table__header"
                  scope="col"
                >
                  Date and time
                </th>
                <th
                  class="table__header"
                  scope="col"
                >
                  Information
                </th>
              </tr>
            </thead>
            <tbody
              class="table__body"
            >
              <tr
                class="table__row"
                data-testid="instrument-table-row"
              >
                <td
                  class="table__cell "
                >
                  19/02/2021 12:46:29
                </td>
                <td
                  class="table__cell "
                >
                  <span
                    class="status status--info"
                  >
                    Successfully uninstalled questionnaire OPN2012K
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumb"
      class="breadcrumb"
    >
      <ol
        class="breadcrumb__items u-fs-s"
      >
        <li
          class="breadcrumb__item"
          id="breadcrumb-0"
        >
          <a
            class="breadcrumb__link"
            href="/"
          >
            Home
          </a>
          <svg
            class="svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
      </ol>
    </nav>
    <main
      class="page__main u-mt-no"
      id="main-content"
    >
      <h1
        class="u-mb-l"
      >
        Questionnaire deployment history
      </h1>
      <button
        class="btn btn--small "
        data-testid="button"
        type="button"
      >
        <span
          class="btn__inner"
        >
          Reload
        </span>
      </button>
      <div
        class="elementToFadeIn"
      >
        <table
          class="table"
          id="audit-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                class="table__header"
                scope="col"
              >
                Date and time
              </th>
              <th
                class="table__header"
                scope="col"
              >
                Information
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body"
          >
            <tr
              class="table__row"
              data-testid="instrument-table-row"
            >
              <td
                class="table__cell "
              >
                19/02/2021 12:46:29
              </td>
              <td
                class="table__cell "
              >
                <span
                  class="status status--info"
                >
                  Successfully uninstalled questionnaire OPN2012K
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
